onnxruntime/contrib_ops/cpu/quantization/qlinear_softmax.cc

```diff
@@ -0,0 +1,373 @@
+            *y_t++ = vy;
+          } while (--elements_n != 0);
+          x_t = x_t_cur;
+        }
+      });
+
+  return Status::OK();
+}
+
+template <>
+common::Status QlinearSoftmaxCPU<int8_t>(size_t N,
+                                         size_t D,
+                                         const int8_t* x_data,
+                                         int8_t* y_data,
+                                         const uint32_t* lookup_table,
+                                         uint32_t y_scale,
+                                         int8_t yzp,
+                                         onnxruntime::concurrency::ThreadPool* thread_pool) {
+  using onnxruntime::TensorOpCost;
+  using onnxruntime::concurrency::ThreadPool;
+  ThreadPool::TryParallelFor(
+      thread_pool, N,
+      TensorOpCost{static_cast<double>(D * 3),
+                   static_cast<double>(D),
+                   static_cast<double>(D * 3)},
+      [x_data, y_data, D, y_scale, yzp, &lookup_table](std::ptrdiff_t first, std::ptrdiff_t last) {
+        const auto c_y_scale = y_scale;
+        const auto c_y_zp = yzp;
+
+        const int8_t* x_t = x_data + first * D;
+        int8_t* y_t = y_data + first * D;
+        for (; first < last; first++) {
+          // reduceMaxUint8
+          int8_t xmax = *std::max_element(x_t, x_t + D);
+          const size_t adjustment = 127 - xmax;
+          const uint32_t* shifted_lookuptable = lookup_table;
+          size_t elements_n = D;
+          // reduceSumUin8ToUint32: need speedup
+          uint32_t vsum = 0;
+          const int8_t* x_t_cur = x_t;
+          do {
+            const size_t vx = uint8_t(adjustment + *x_t_cur++);
+            vsum += shifted_lookuptable[vx];
+          } while (--elements_n != 0);
+          if (vsum == 0) {
+            return;
+          }
+          elements_n = D;
+          x_t_cur = x_t;
+          // elementwise div
+          const uint32_t vrounding = (vsum >> 1);
+          do {
+            const size_t vx = uint8_t(adjustment + *x_t_cur++);
+            const uint32_t vt = shifted_lookuptable[vx];
+            // simulate round function, and re-quant to int8
+            const uint32_t vq = ((vt * c_y_scale) + vrounding) / vsum + c_y_zp;
+            const int8_t vy = static_cast<int32_t>(vq) > 255 ? static_cast<int8_t>(255) : static_cast<int8_t>(vq);
+            *y_t++ = vy;
+          } while (--elements_n != 0);
+          x_t = x_t_cur;
+        }
+      });
+
+  return Status::OK();
+}
+
+template <typename T>
+const uint32_t* QLinearSoftmax<T>::GetLookupTable(OpKernelContext* context, size_t reduce_len) const {
+  const uint32_t* lookup_table = fixed_lookup_table_.data();
+  if (fixed_lookup_table_.size() == 0) {
+    tmp_lookup_table_.resize(256);
+    lookup_table = tmp_lookup_table_.data();
+    const float X_scale = *(context->Input<Tensor>(1)->Data<float>());
+    QlinearBuildLookupTableUint32<T>(tmp_lookup_table_.data(), X_scale, reduce_len);
```

Comment:
you do an allocation in tmp_lookup_table_ (so tmp_lookup_table_ owns that data) and assign it to the pointer in lookup_table (which is a local variable). what actually a) allocates data in the fixed_lookup_table_ vector, and b) transfers the values in tmp_lookup_table_ to it? 

Possibly a moot question if the above suggestion to rework how the lookup table is created works.

Reply:
What I think is actually happening is that you're re-creating the lookup table in tmp_lookup_table_ on every call to GetLookupTable given tmp_lookup_table_.data() is the pointer being returned. Guessing that would break (silently) as soon as you had concurrent calls given tmp_lookup_table_ could have a `resize` done from a second call whilst the first call was still using the data.

Reply:
Thanks for your pointing out.  Would one object be running in multiple thread? I don't know that.

I don't want to assign data in `fixed_lookup_table_` for its meaning is straightforward, quantized paraments is fixed.  Then I have to re-calculate the lookup table every time if it's dynamic. And I am worried about the memory alloc cost.

If Op has to be keep thread-safe. I would rather alloc 256x4bytes on the stack, each thread owns a seperate memory.